rts/Sim/Misc/YardmapStatusEffectsMap.h

```diff
@@ -0,0 +1,149 @@
+    uint8_t& GetMapStateNative(int x, int z) { return stateMap[interleave(x, z)]; }
+
+    bool AreAllFlagsSet(int x, int z, uint8_t flags) { return (GetMapState(x, z) & flags) == flags; }
+	bool AreAnyFlagsSet(int x, int z, uint8_t flags) { return (GetMapState(x, z) & flags) != 0; }
+	
+    void SetFlags(int x, int z, uint8_t flags) { GetMapState(x, z) |= flags; }
+    void ClearFlags(int x, int z, uint8_t flags) { GetMapState(x, z) &= ~flags; }
+
+    void InitNewExitOnlyMap();
+
+    typedef std::vector<uint8_t> ExitOnlyMapType;
+    ExitOnlyMapType stateMap;
+};
+
+extern YardmapStatusEffectsMap exitOnlyMap;
+
+struct ObjectCollisionMapHelper {
+	bool collidesUnderWater;
+	bool collidesAboveWater;
+
+	ObjectCollisionMapHelper(const CSolidObject& object) {
+		SetObjectCollisionStates(object);
+	}
+
+	ObjectCollisionMapHelper(const MoveDef& moveDef, float ypos) {
+		SetMoveDefCollisionStates(moveDef, ypos);
+	}
+
+	ObjectCollisionMapHelper(const MoveDef& moveDef) {
+		SetMoveDefCollisionStates(moveDef);
+	}
+
+	float GetMoveCollisionHeight(const CSolidObject& object) const {
+		if (object.moveDef != nullptr)
+			return object.moveDef->height;
+		else
+			return object.height;
+	}
+
+	bool IsOnWaterSurface(const CSolidObject& object) const {
+		if (object.moveDef != nullptr)
+			return !object.moveDef->isSubmersible;
+		else {
+			auto unit = dynamic_cast<const CUnit*>(&object);
+			if (unit != nullptr)
+				return unit->FloatOnWater();
+		}
+		return false;
+	}
+
+	void SetObjectCollisionStates(const CSolidObject& object) {
+		const bool floatsOnWater = IsOnWaterSurface(object);
+		collidesUnderWater = !floatsOnWater;
+		collidesAboveWater = (floatsOnWater || object.pos.y + GetMoveCollisionHeight(object) >= 0.f);
+	}
+
+    void SetMoveDefCollisionStates(const MoveDef& moveDef, float ypos) {
+		const bool floatsOnWater = !moveDef.isSubmersible;
+		collidesUnderWater = !floatsOnWater;
+		collidesAboveWater = (floatsOnWater || ypos + moveDef.height >= 0.f);
+    }
+
+    void SetMoveDefCollisionStates(const MoveDef& moveDef) {
+		collidesUnderWater = moveDef.isSubmersible;
+		collidesAboveWater = true;
+    }
+
+	uint8_t GetExitOnlyFlags() const {
+		return (YardmapStatusEffectsMap::EXIT_ONLY_BW * collidesUnderWater)
+			+ (YardmapStatusEffectsMap::EXIT_ONLY_AW * collidesAboveWater);
+	}
+
+    bool IsExitOnlyAt(int x, int z) const {
+		return exitOnlyMap.AreAllFlagsSet(x, z, GetExitOnlyFlags());
+	}
+
+	void SetExitOnlyAt(int x, int z) const {
+		exitOnlyMap.SetFlags(x, z, GetExitOnlyFlags());
+	}
+
+	void ClearExitOnlyAt(int x, int z) const {
+		exitOnlyMap.ClearFlags(x, z, GetExitOnlyFlags());
+	}
+
+	uint8_t GetBlockBuildingFlags() const {
+		return (YardmapStatusEffectsMap::BLOCK_BUILDING_BW * collidesUnderWater)
+			+ (YardmapStatusEffectsMap::BLOCK_BUILDING_AW * collidesAboveWater);
+	}
+
+    bool IsBlockBuildingAt(int x, int z) const {
```

Comment:
a bunch of bad ws (tabs/spaces mix) in the file

Reply:
Out of curiosity, what is the engine expected settings with respect to tabs and spaces?